gossip/src/crds_value.rs

```diff
@@ -490,84 +493,213 @@ impl Sanitize for NodeInstance {
     }
 }
 
-#[derive(Serialize, Deserialize, Clone, Default, PartialEq, Eq, AbiExample, Debug)]
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, AbiExample, AbiEnumVisitor)]
+pub enum SlotsOffsets {
+    RunLengthEncoding(RunLengthEncoding),
+    RawOffsets(RawOffsets),
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq, AbiExample)]
+struct U16(#[serde(with = "serde_varint")] u16);
+
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq, AbiExample)]
+pub struct RunLengthEncoding {
+    // The vector always starts with 1. Encode number of 1's and 0's consecutively.
+    // For example, 110000111 is [2, 4, 3].
+    encoded: Vec<U16>,
+}
+
+impl RunLengthEncoding {
+    // Return whether the encoded vec is already full.
+    fn add_number_and_test_full(
+        encoded: &mut Vec<U16>,
+        total_serde_bytes: &mut usize,
+        current_bit_count: u16,
+    ) -> bool {
+        if current_bit_count == 0 {
+            false
+        } else {
+            let serde_bits = 16 - current_bit_count.leading_zeros();
+            let serde_bytes = ((serde_bits + 6) / 7).max(1) as usize;
+            if *total_serde_bytes + serde_bytes > RestartLastVotedForkSlots::MAX_SPACE {
+                true
+            } else {
+                encoded.push(U16(current_bit_count));
+                *total_serde_bytes += serde_bytes;
+                false
+            }
+        }
+    }
+
+    pub fn new(bits: &BitVec<u8>) -> Self {
+        let mut encoded: Vec<U16> = Vec::new();
+        let mut current_bit = 1;
+        let mut current_bit_count: u16 = 0;
+        let mut total_serde_bytes = 0;
+        for i in 0..bits.len() {
+            let bit: u8 = bits.get(i) as u8;
+            if bit == current_bit {
+                current_bit_count += 1;
+            } else {
+                if Self::add_number_and_test_full(
+                    &mut encoded,
+                    &mut total_serde_bytes,
+                    current_bit_count,
+                ) {
+                    current_bit_count = 0;
+                    break;
+                }
+                current_bit = bit;
+                current_bit_count = 1;
+            }
+        }
+        Self::add_number_and_test_full(&mut encoded, &mut total_serde_bytes, current_bit_count);
+        Self { encoded }
+    }
+
+    pub fn slots_count(&self) -> usize {
+        self.encoded.iter().map(|x| x.0 as usize).sum::<usize>()
+    }
+
+    pub fn to_slots(&self, last_slot: Slot, min_slot: Slot) -> Vec<Slot> {
+        let mut result: Vec<Slot> = Vec::new();
+        let mut offset = 0;
+        let mut current_bit = 1;
+        for bit_count in &self.encoded {
```

Comment:
instead of manually maintaining current_bit
```rust
self.encoded.iter().zip([1, 0].iter().cycle())
```

https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.cycle

Reply:
Changed.